llvm/docs/SYCLOffloadingDesign.rst

```diff
@@ -0,0 +1,677 @@
+which:
+1. Assigns numeric IDs to specialization constants found in the linked module.
+2. Transforms IR to a form expected by the LLVM to SPIR-V translation step.
+3. Translates SPIR-V friendly IR to SPIR-V IR during LLVM to SPIR-V
+transation step.
+1. Collects and provides \<Symbolic ID\> =\> \<numeric IDs + additional info\>
+mapping, which is later being used by SYCL runtime to set specialization
+constant values provided by user.
+
+Remainder of this section explains these three steps in detail.
+
+Assignment of numeric IDs to specialization constants
+`````````````````````````````````````````````````````
+
+This task is achieved by maintaining a map, which holds a list of numeric IDs
+for each encountered symbolic ID of a specialization constant. Those IDs are
+used to identify the specialization constants at SPIR-V level.
+
+As noted above one symbolic ID can have several numeric IDs assigned to it -
+such 1:N mapping comes from the fact that at SPIR-V level, composite
+specialization constants don't have dedicated IDs and they are being identified
+and specialized through their scalar leafs and corresponding numeric IDs.
+
+For example, the following code:
+
+  .. code-block:: c++
+    :linenos:
+
+      struct Nested {
+        float a, b;
+      };
+      struct A {
+        int x;
+        Nested n;
+      };
+      constexpr specialization_id<int> id_int;
+      constexpr specialization_id<A> id_A;
+      // ...
+        [=](kernel_handler h) {
+          h.get_specialization_constant<id_int>();
+          h.get_specialization_constant<id_A>();
+        }
+
+Will result in the following numeric IDs assignment:
+
+  .. code-block:: c++
+    :linenos:
+
+      // since `id_int` is a simple arithmetic specialization constant, we only
+      // have a single numeric ID associated with its symbolic ID
+      unique_symbolic_id_for_id_int -> { 0 }
+      // `id_A` is a composite with three leafs (scalar members, including ones
+      // located in nested composite types), which results in three numeric IDs
+      // associated with the same symbolic ID
+      unique_symbolic_id_for_id_A -> { 1, 2, 3 }
+
+As it is shown in the example above, if a composite specialization constant
+contains another composite within it, that nested composite is also being
+"flattened" and its leafs are considered to be leafs of the parent
+specialization constants. This done by depth-first search through the composite
+elements.
+
+Transformation of LLVM IR to SPIR-V friendly IR form
+````````````````````````````````````````````````````
+
+SPIR-V friendly IR form is a special representation of LLVM IR, where some
+function are named in particular way in order to be recognizable by the SPIR-V
+translator to convert them into corresponding SPIR-V instructions later.
+The format is documented [here][spirv-friendly-ir] `spirv-friendly-ir <https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/docs/SPIRVRepresentationInLLVM.rst>`_
```

Comment:
Are we planning to upstream this support? My understanding is that the new LLVM-IR replacements are in development.